File: 0692-top-k-frequent-words/0692-top-k-frequent-words.py
class Solution(object):
    def topKFrequent(self, words, k):
        """
        :type words: List[str]
        :type k: int
        :rtype: List[str]
        """
        counter = Counter(words)
        heap = [(-freq, word) for word, freq in counter.items()]
        heapify(heap)

        topKFreq = []

        for i in range(k):
            topKFreq.append(heappop(heap)[1])
        
        return topKFreq


        

File: 0692-top-k-frequent-words/0692-top-k-frequent-words.cpp
class Solution {
public:
    vector<string> topKFrequent(vector<string>& words, int k) {
        unordered_map<string, int> cnt;
        vector<pair<int, string>> h;
        vector<string> res;
        for (string& word : words) {
            cnt[word]++;
        }
        for (auto& p : cnt) {
            h.push_back({-p.second, p.first});
        }

        make_heap(h.begin(), h.end(), greater<pair<int, string>>());
        for (int i = 0; i < k; i++) {
            pop_heap(h.begin(), h.end(), greater<pair<int, string>>());
            res.push_back(h.back().second);
            h.pop_back();
        }
        return res;
    }
};